### PR TITLE
Added Box2d Lua module

### LINKED ISSUE #5405: Expose allow_sleep and awake Box2D parameters
**Is your feature request related to a problem? Please describe (REQUIRED):**
Dynamic bodies are deactivated from the physics simulation if no forces have been affecting them for a certain time.

In my zero gravity game the player starts the game by connecting the player body to a different body. If the player body has gone to sleep before creating the joint, the joint doesn't work - it connects but is passive. This means physics.set_joint_properties() no longer has any effect.

**Describe the solution you'd like (REQUIRED):**
Expose the allow_sleep flag in Box2D so that that a body can be set to never go to sleep. Also add an physics.is_awake() to check if a body is sleeping, if using allow_sleep is not desired.

**Describe alternatives you've considered (REQUIRED):**
Applying a small impulse every frame to keep the body awake.

**Additional context (OPTIONAL):**
```
b2BodyDef bodyDef;
bodyDef.allowSleep = true;
bodyDef.awake = true;
```
https://box2d.org/documentation/md__d_1__git_hub_box2d_docs_dynamics.html

### LINKED ISSUE #8256: Add apply impulse to Defold Physics 2D/3D API
**Is your feature request related to a problem? Please describe (REQUIRED):**
I’m trying to port a Unity game to Defold, it relies heavily on the 2D physics engine, which fortunately is Box2D in both.

In our Unity games, we use a lot [ForceMode2D.Impulse](https://docs.unity3d.com/ScriptReference/ForceMode2D.Impulse.html), which I guess translates internally to [ApplyLinearImpulse](https://box2d.org/documentation/classb2_body.html#a7f677e93efb3c4c065087aff317274a3)

But in Defold, I only see [apply_force](https://defold.com/ref/physics/#apply_force)

**Describe the solution you'd like (REQUIRED):**
Same way as we currently have [apply_force](https://defold.com/ref/physics/#apply_force), I'd like to have `apply_impulse`
Internally, for 2D it would call [ApplyLinearImpulse](https://box2d.org/documentation/classb2_body.html#a7f677e93efb3c4c065087aff317274a3)
for 3D, Bullet has [applyImpulse](https://github.com/bulletphysics/bullet3/blob/39b8de74df93721add193e5b3d9ebee579faebf8/src/BulletDynamics/Dynamics/btRigidBody.h#L335), but I can't find it in the code source of Defold (not sure if maybe it's because the actual source are not in the repo, but downloaded later when building ?)
Or like Unity did, add a parameter to the apply_force to specify if it's a force or an impulse.

I'd be happy to contribute a potential solution if we can agree on something.

**Describe alternatives you've considered (REQUIRED):**
I believe I could maybe set [linear_velocity](https://defold.com/ref/physics/#linear_velocity) taking into account the mass of my rigidbody, but I’d rather let the physics engine set the final velocity of the rigidbodies.

Alternatively, I could just bring Box2D directly via https://defold.com/assets/defold-box2d/, but it seems it would require much more code and time to then use physics in the game, which relies heavily on physics to begin with...

**Additional context (OPTIONAL):**
* I mentioned this problem initially on Defold forum, cf. https://forum.defold.com/t/apply-impulse-in-2d-physics/74388
* The game I'm trying to port: https://surmoi.itch.io/baijiu-drifters

### PR DESCRIPTION
**Pull request by JCash**
This allows you to get a physics body and manipulate its forces, velocities and other properties.

Example:
```lua
local id = factory.create("#factory", position)
local url = msg.url(nil, id, "collisionobject") -- get the collision object in the spawned object
self.body = b2d.get_body(url) -- get the b2body object
b2d.body.dump(self.body) -- debug print

-- add an impulse
local pos = b2d.body.get_position(self.body)
b2d.body.apply_linear_impulse(self.body, vmath.vector3(300,200,0), pos + vmath.vector3(16,16,0))
```

A full list of functions is available in the reference api for `b2d` and `b2d.body`.
Note that this was a first step. Next up is adding the missing structs, e.g. world, joints, fixtures and shapes support.

Fixes https://github.com/defold/defold/issues/8256
Fixes https://github.com/defold/defold/issues/5405
Fixes https://github.com/defold/defold/issues/3191
Fixes https://github.com/defold/defold/issues/3120

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/engine/src/engine.cpp [hunk 1]
```diff
@@ -1305,6 +1305,7 @@ namespace dmEngine
         script_lib_context.m_HidContext      = engine->m_HidContext;
         script_lib_context.m_GraphicsContext = engine->m_GraphicsContext;
         script_lib_context.m_JobThread       = engine->m_JobThreadContext;
+        script_lib_context.m_ConfigFile      = engine->m_Config;
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1308]: To allow for reading game properties, like `physics.scale`

File: engine/gamesys/src/gamesys/scripts/script_box2d.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    dmVMath::Vector3 FromB2(const b2Vec2& p, float inv_scale)
+    {
+        return dmVMath::Vector3(p.x * inv_scale, p.y * inv_scale, 0);
+    }
+
+    void PushWorld(struct lua_State* L, class b2World* world)
+    {
+        lua_pushlightuserdata(L, world);
+    }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d.cpp:71]: todo: add script_box2d_world.cpp with proper user type.

File: engine/gamesys/src/gamesys/scripts/script_box2d.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    //////////////////////////////////////////////////////////////////////////////
+
+    static void GetCollisionObject(lua_State* L, int index, dmGameObject::HCollection collection, void** comp, void** comp_world)
+    {
+        dmGameObject::GetComponentUserDataFromLua(L, index, collection, COLLISION_OBJECT_EXT, (uintptr_t*)comp, 0, comp_world);
+    }
+
+    static int B2D_GetWorld(lua_State* L)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d.cpp:96]: These parts should be moved to a world specific .cpp file

File: engine/gamesys/src/gamesys/scripts/script_box2d.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+#define SETCONSTANT(NS, NAME) \
+        lua_pushnumber(L, (lua_Number) NS :: NAME); \
+        lua_setfield(L, -2, #NAME);
+
+        lua_newtable(L);
+        SETCONSTANT(b2BodyType, b2_staticBody);
+        SETCONSTANT(b2BodyType, b2_kinematicBody);
+        SETCONSTANT(b2BodyType, b2_dynamicBody);
+        lua_setfield(L, -2, "b2BodyType");
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d.cpp:156]: They'll show up as `box2d.b2bodyType.b2_staticBody`

File: engine/physics/src/physics/physics.h [hunk 5]
```diff
@@ -347,6 +347,9 @@ namespace dmPhysics
      */
     void DeleteWorld2D(HContext2D context, HWorld2D world);
 
+    void* GetWorldContext2D(HWorld2D world);
+    void* GetCollisionObjectContext2D(HCollisionObject2D collision_object);
```
Comment (JCash) [inline engine/physics/src/physics/physics.h:351]: We need to resolve the glue between our physics engine and box2d

File: engine/physics/src/wscript [hunk 6]
```diff
@@ -7,8 +8,27 @@ def configure(conf):
     conf.recurse('box2d')
     conf.recurse('physics')
 
+def dmsdk_add_header_files(bld, target, source):
```
Comment (JCash) [inline engine/physics/src/wscript:11]: Copy Box2d headers to our internal sdk

File: engine/script/src/script.cpp [hunk 7]
```diff
@@ -698,6 +698,23 @@ namespace dmScript
         return type_hash;
     }
 
+
+    uint32_t RegisterUserTypeLocal(lua_State* L, const char* name, const luaL_reg meta[])
```
Comment (JCash) [inline engine/script/src/script.cpp:702]: I opted for a separate function to keep the RegisterUserType() clean.
This function allows us to register an "object style" syntax for the b2d object.
I felt that that was cleaner, and easier to browse the api between the official Box2D docs and also to understand how our module is working.

File: engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp [hunk 8]
```diff
@@ -0,0 +1,885 @@
...
+    static b2Body* ToBody(lua_State* L, int index)
+    {
+        return (b2Body*)dmScript::ToUserType(L, index, TYPE_HASH_BODY);
+    }
+
+    static int Body_GetPosition(lua_State* L)
+    {
+        DM_LUA_STACK_CHECK(L, 1);
+        b2Body* body = CheckBody(L, 1);
+        dmScript::PushVector3(L, FromB2(body->GetPosition(), GetInvPhysicsScale()));
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp:66]: Just like in our old implementation, we need to take take of conversion between units, using FromB2/CheckVec2.

File: engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp [hunk 9]
```diff
@@ -0,0 +1,885 @@
...
+        return 1;
+    }
+
+    static int Body_SetTransform(lua_State* L)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp:70]: GetTransform currently left out as I wasn't sure what to return. 2-tuple?

File: engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp [hunk 10]
```diff
@@ -0,0 +1,885 @@
...
+    static const luaL_reg Body_meta[] =
+    {
+        {"__tostring", Body_tostring},
+        {"__eq", Body_eq},
+
+        {"get_position", Body_GetPosition},
+        {"set_transform", Body_SetTransform},
+
+        //{"get_user_data", Body_GetUserData}, - could return the game object id ?
+        //{"set_user_data", Body_SetUserData}, - could attach the body to a game object?
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp:455]: Getting the id of the game object is useful, and it should be possible to set the id as well, with some extra component functions to make sure that the game+physics objects life times are considered.

File: engine/gamesys/src/gamesys/scripts/script_box2d.cpp [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+        dmGameObject::HCollection collection = dmGameObject::GetCollection(CheckGoInstance(L));
+
+        void* component = 0;
+        GetCollisionObject(L, 1, collection, &component, 0);
+
+        b2Body* body = dmGameSystem::CompCollisionObjectGetBox2DBody(component);
+
+        if (body)
+            PushBody(L, body);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d.cpp:124]: I need to make this more robust, as it currently contains an internal pointer, and the game object may be deleted directly after (and along with is the collision object).

E.g. if this has a game object id set, then we need to check it before doing any operations on the body.

File: engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp [hunk 12]
```diff
@@ -0,0 +1,885 @@
...
+    {
+        {0,0}
+    };
+
+    static const luaL_reg Body_meta[] =
+    {
+        {"__tostring", Body_tostring},
+        {"__eq", Body_eq},
+
+        {"get_position", Body_GetPosition},
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp:451]: The previous extension, uses the Box2D naming style: `GetPosition()`.
Here I'm using the naming style in Defold.

Yay, or nay?

File: engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp [hunk 13]
```diff
@@ -0,0 +1,885 @@
...
+
+        {"get_mass", Body_GetMass},
+        {"get_inertia", Body_GetInertia},
+        {"get_angle", Body_GetAngle},
+
+        // {"get_mass_data", Body_GetMassData},
+        // {"set_mass_data", Body_SetMassData},
+        // {"reset_mass_data", Body_ResetMassData},
+        // {"synchronize_fixtures", SynchronizeFixtures},
+        // SynchronizeSingle(b2Shape* shape, int32 index)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp:465]: For mvp2

File: engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp [hunk 14]
```diff
@@ -0,0 +1,885 @@
...
+
+/*# Box2D b2Body documentation
+ *
+ * Functions for interacting with Box2D bodies.
+ *
+ * @document
+ * @name b2body
+ */
+
+/**
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_box2d_body.cpp:544]: Copied all documentation, but set some to "undocumented" `/**` until they're implemented.

File: engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp [hunk 15]
```diff
@@ -0,0 +1,971 @@
...
+ * @warning This function is locked during callbacks.
+ * @name body:destroy_fixture
+ * @param fixture [type: b2Fixture] the world position of the body's origin.
+ */
+
+/*#
+ * Set the position of the body's origin and rotation.
+ * This breaks any contacts and wakes the other bodies.
+ * Manipulating a body's transform may cause non-physical behavior.
+ * @name body:set_transform
```
Comment (AGulev) [inline engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp:667]: documentation still uses `body:*` syntax

File: engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp [hunk 16]
```diff
@@ -0,0 +1,971 @@
...
+ */
+
+/*# Get the next body in the world's body list.
+ * @name body:get_next
+ * @return body [type: b2Body] the next body
+ */
+
+/*# Get the user data pointer that was provided in the body definition.
+ * @name body:get_user_data
+ * @return id [type: hash] the game object id this body is connected to
```
Comment (AGulev) [inline engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp:948]: this function isn't implemented, should be hidden from docs

File: engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp [hunk 17]
```diff
@@ -0,0 +1,971 @@
...
+ */
+
+/*# Print the body representation to the log output
+ * @name body:dump
+ */
+
+/** Get the total force currently applied on this object
+ * @name body:get_force
+ * @note Defold Specific
+ * @return force [type: vmath.vector3]
```
Comment (AGulev) [inline engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp:968]: it seems like this function is already implemented but it's commented (as I understand `/**` - comment out function, and `/*#` - added function?)

**Review by AGulev — DISMISSED** [review]
I like it a lot!

A couple of small things should be fixed in docs.
+ tests are still red.

**Review by Jhonnyg — DISMISSED** [review]

File: engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp (hunk 17 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/scripts/box2d/script_box2d_body.cpp:968]: Correct.

File: engine/gamesys/src/gamesys/scripts/box2d/script_box2d.cpp [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+//         SETCONSTANT(b2JointType, e_pulleyJoint)
+//         SETCONSTANT(b2JointType, e_mouseJoint)
+//         SETCONSTANT(b2JointType, e_gearJoint)
+//         SETCONSTANT(b2JointType, e_wheelJoint)
+//         SETCONSTANT(b2JointType, e_weldJoint)
+//         SETCONSTANT(b2JointType, e_frictionJoint)
+//         SETCONSTANT(b2JointType, e_ropeJoint)
+//         lua_setfield(L, -2, "b2JointType");
+
+// #undef SETCONSTANT
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/gamesys/src/gamesys/scripts/box2d/script_box2d.cpp:163]: Kind of both. I'll remove it.

**Pull request merged**